src/ucp/core/ucp_worker.c

```diff
@@ -1318,7 +1318,71 @@ static void ucp_worker_init_cpu_atomics(ucp_worker_h worker)
     }
 }
 
-static void ucp_worker_init_device_atomics(ucp_worker_h worker)
+static int ucp_resources_is_device_shared(const ucp_tl_resource_desc_t *rsc1,
+                                          const ucp_tl_resource_desc_t *rsc2)
+{
+    return (rsc1->md_index == rsc2->md_index) &&
+           !strncmp(rsc1->tl_rsc.dev_name, rsc2->tl_rsc.dev_name,
+                    UCT_DEVICE_NAME_MAX);
+}
+
+static ucs_status_t
+ucp_worker_loopback_amo_ep_init(ucp_worker_h worker,
+                                const ucp_tl_resource_desc_t *best_rsc)
+{
+    ucp_context_h context = worker->context;
+    uint64_t tl_bitmap    = worker->atomic_tls;
+    ucp_tl_resource_desc_t *rsc;
+    ucp_rsc_index_t rsc_index;
+
+    /*
+     * NOTE: worker->atomic_tls can contain more than 1 transport with AMO
+     *       capabilities to the same device and/or no one EP2IFACE for wireup
+     *       needs.
+     *       If there are p2p only TLS, need to add other TLs for wireup.
+     *       If there are at leat 1 EP2IFACE transport with AMO capabilities, we
+     *       can optimize consumption of resources reducing bitmap to 1 TL.
+     */
+
+    ucs_for_each_bit(rsc_index, worker->atomic_tls) {
+        if (ucp_worker_is_tl_p2p(worker, rsc_index)) {
+            /* disable p2p TLS to check if we need AUX TLS */
+            tl_bitmap &= ~UCS_BIT(rsc_index);
+        }
+    }
+
+    if (tl_bitmap == 0) {
+        /* only p2p, need to enable all TLS on AMO device for wireup */
+        ucs_for_each_bit(rsc_index, context->tl_bitmap) {
+            rsc = &context->tl_rscs[rsc_index];
+            if (ucp_resources_is_device_shared(rsc, best_rsc)) {
+                tl_bitmap |= UCS_BIT(rsc_index);
+            }
+        }
+    } else {
+        ucs_for_each_bit(rsc_index, tl_bitmap) {
+            rsc = &context->tl_rscs[rsc_index];
+            if (rsc == best_rsc) {
```

Comment:
why do we need to select best resource here ?
isn't it already done by wireup logic on the loopback ep

Reply:
it is but intention was to reduce resource consumption, see https://github.com/openucx/ucx/pull/4596/files#r361617464